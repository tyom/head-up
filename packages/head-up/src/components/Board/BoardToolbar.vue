<template>
  <div class="BoardToolbar">
    <form class="toolbar-container" @submit.prevent="handleSubmit">
      <div class="toolbar-item _info">
        <input
          v-if="isEditable"
          v-model="boardTitle"
          type="text"
          class="toolbar-input"
          placeholder="Board title"
        >
        <span v-else class="toolbar-title"> {{ title }} (read-only) </span>
      </div>
      <div class="toolbar-item _actions">
        <button v-if="isEditable" :disabled="!hasChanged">Apply</button>
        <button @click.prevent="handleEditDone">OK</button>
      </div>
    </form>
  </div>
</template>

<script>
export default {
  props: {
    title: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      boardTitle: this.title,
    };
  },
  computed: {
    hasChanged() {
      return this.boardTitle !== this.title;
    },
    isEditable() {
      return this.$parent.editable;
    },
  },
  inject: ['handleEditDone'],
  methods: {
    handleSubmit() {
      this.$emit('save', {
        title: this.boardTitle,
      });
    },
  },
};
</script>

<style scoped>
.BoardToolbar {
  background: radial-gradient(
      circle at top left,
      rgba(#20416d, 0.3),
      rgba(#20416d, 0.6)
    )
    no-repeat;
  height: 3.5rem;
  display: flex;
  align-items: stretch;
  position: relative;
}

.toolbar-container {
  flex: 1;
  display: flex;
  align-items: stretch;
}

.toolbar-item {
  display: flex;
  align-items: center;
  padding: 0.6rem 0.5em;
}

.toolbar-title {
  font-size: 1rem;
  padding: 0.5rem;
}

.toolbar-input {
  background: rgba(#fff, 0.1);
  color: #fff;
  border-radius: 0.3rem;
  padding: 0.5rem;
  min-width: 10em;
  width: 100%;
  border: 0;
  font-size: 1rem;
  text-align: inherit;
  outline: 0;
  transition: 0.2s;

  &:hover {
    background: rgba(#fff, 0.15);
  }

  &:focus {
    color: #000;
    background-color: #fff6;
    box-shadow: inset 0 0 2px 1px #0004;
  }

  &::placeholder {
    color: #fff4;
  }

  @media (min-width: 50em) {
    max-width: 50%;
  }
}

.toolbar-item._info {
  flex: 1;
}

.toolbar-item._actions {
  margin-left: auto;

  & button {
    border-radius: 0.3rem;
    padding: 0.6rem 1em;
    font-size: 0.9rem;
    line-height: 1;
    background: #0005;
    border: 1px solid #fff3;
    color: #fff;
    outline: 0;

    &:hover {
      border-color: #fff5;
    }

    &:focus {
      border-color: #fff8;
    }

    &:active {
      background: #000;
    }

    &[disabled] {
      opacity: 0.4;
      cursor: default;
    }
  }

  & button + button {
    margin-left: 0.5em;
  }
}
</style>
