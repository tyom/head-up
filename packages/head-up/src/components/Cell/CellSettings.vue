<template>
  <div class="CellSettings">
    <h2>Settings</h2>
  </div>
</template>

<style scoped>
.CellSettings {
  padding: 1rem;
  position: absolute;
  background-color: #000e;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}
</style>
