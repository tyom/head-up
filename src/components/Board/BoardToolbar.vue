<template>
  <div class="BoardToolbar">
    <form class="toolbar-container" @submit.prevent="$emit('save', boardState)">
      <div class="toolbar-item _info">
        <input
          v-if="boardState.editable"
          v-model="boardState.title"
          type="text"
          class="toolbar-input"
          placeholder="Board title"
        />
        <span v-else class="toolbar-title">
          {{ boardState.title }} (read-only)
        </span>
      </div>
      <div class="toolbar-item _actions">
        <button v-if="boardState.editable" :disabled="!hasChanged" class="save">
          Apply
        </button>
        <button class="done" @click.prevent="$emit('done', boardState)">
          OK
        </button>
      </div>
    </form>
  </div>
</template>

<script>
export default {
  props: {
    board: {
      type: Object,
      default: () => {},
    },
  },
  data() {
    return {
      boardState: { ...this.board },
    };
  },
  computed: {
    hasChanged() {
      return this.boardState.title !== this.board.title;
    },
  },
};
</script>

<style scoped>
.BoardToolbar {
  background: radial-gradient(
      circle at top left,
      rgba(#20416d, 0.3),
      rgba(#20416d, 0.6)
    )
    no-repeat;
  height: 3.5rem;
  display: flex;
  align-items: stretch;
  position: relative;
}

.toolbar-container {
  flex: 1;
  display: flex;
  align-items: stretch;
}

.toolbar-item {
  display: flex;
  align-items: center;
  padding: 0.6rem 0.5em;
}

.toolbar-title {
  font-size: 1rem;
  padding: 0.5rem;
}

.toolbar-input {
  background: rgba(#fff, 0.1);
  color: #fff;
  border-radius: 0.3rem;
  padding: 0.5rem;
  min-width: 10em;
  width: 100%;
  border: 0;
  font-size: 1rem;
  text-align: inherit;
  outline: 0;
  transition: 0.2s;

  &:hover {
    background: rgba(#fff, 0.15);
  }

  &:focus {
    color: #000;
    background-color: #fff6;
    box-shadow: inset 0 0 1px 1px #0006, 0 0 0 1px #fffa;
  }

  &::placeholder {
    color: #fff4;
  }

  @media (min-width: 50em) {
    max-width: 50%;
  }
}

.toolbar-item._info {
  flex: 1;
}

.toolbar-item._actions {
  margin-left: auto;

  & button {
    border-radius: 0.3rem;
    padding: 0.6rem 1em;
    font-size: 0.9rem;
    line-height: 1;
    background: #0005;
    border: 1px solid #fff3;
    color: #fff;
    outline: 0;

    &:hover {
      border-color: #fff5;
    }

    &:focus {
      border-color: #fff8;
    }

    &:active {
      background: #000;
    }

    &[disabled] {
      opacity: 0.4;
      cursor: default;
    }
  }

  & button + button {
    margin-left: 0.5em;
  }
}
</style>
