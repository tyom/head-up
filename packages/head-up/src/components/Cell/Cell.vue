<template>
  <div class="Cell">
    <header v-if="isEditable || title" class="header">
      <h2 class="title">{{ title }}</h2>
      <button
        v-if="isEditable"
        :class="{_toggled: showSettings}"
        class="header-button"
        @click.stop="handleSettingsClick"
      >
        <v-icon name="ellipsis-h"/>
      </button>
    </header>
    <div class="container">
      <div :class="bodyClass" class="body">
        <slot v-if="$slots.default"/>
        <template v-else>
          <component
            v-for="(item, idx) in content"
            :is="item.type"
            :key="idx"
            v-bind="item.props"
            class="body-item"
          />
        </template>
      </div>
      <transition name="revealSettings">
        <CellSettings v-if="isEditable && showSettings"/>
      </transition>
    </div>
  </div>
</template>

<script>
import CellSettings from './CellSettings';

export default {
  name: 'Cell',
  components: {
    CellSettings,
  },
  props: {
    title: {
      type: String,
      default: '',
    },
    content: {
      type: Array,
      default: () => [],
    },
    padded: {
      type: Boolean,
      default: false,
    },
    editable: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      showSettings: false,
    };
  },
  computed: {
    isEditable() {
      return this.editable && this.isEditing();
    },
    bodyClass() {
      const content = this.$slots.default ? this.$slots.default : this.content;
      return {
        [`u-grid-${content.length}-x`]: !!content.length,
        padded: this.padded,
      };
    },
  },
  inject: ['isEditing'],
  methods: {
    handleSettingsClick() {
      this.showSettings = !this.showSettings;
    },
  },
};
</script>

<style scoped>
.Cell {
  border: 1px solid #0005;
  transition: 0.3s;
  display: flex;
  flex-direction: column;
  user-select: none;

  &:hover {
    transition: 0.1s;
  }
}

.header {
  font-size: 0.55em;
  color: #fff6;
  position: relative;
  text-transform: uppercase;
  background-color: #0003;
  padding: 0.7em 1em;
  line-height: 1;
  justify-content: center;
  display: flex;
  align-items: center;
}

.header-button {
  border: 0;
  padding: 0.5em 1em;
  background: none;
  color: #fff;
  position: absolute;
  z-index: 1;
  top: 0;
  right: 0;
  bottom: 0;
  opacity: 0.7;
  outline: 0;

  &:hover {
    opacity: 1;
  }

  &:focus-visible {
    background-color: rgba(#20416d, 0.7);
    box-shadow: 0 0 0 1px #fff;
  }

  &._toggled {
    opacity: 1;
    background-color: #20416d;
  }

  &:active {
    opacity: 1;
    transition: none;
    background-color: #20416d;
  }
}

.title {
  font-size: inherit;
}

.container {
  position: relative;
  flex: 1;
  display: flex;
}

.body {
  grid-gap: 1px;
  background-color: rgba(#000, 0.2);
  flex: 1;
  justify-content: stretch;
  align-items: stretch;
}

.body > * {
  background-color: rgba(#fff, 0.03);
}

.padded > * {
  padding: 0.5em;
}

.revealSettings-enter-active,
.revealSettings-leave-active {
  transition: 0.3s;
}
.revealSettings-enter,
.revealSettings-leave-to {
  opacity: 0;
}
</style>
