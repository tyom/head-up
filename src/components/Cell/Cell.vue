<template>
  <div class="Cell">
    <header v-if="isEditable || title" class="header">
      <h2 class="title">{{ title }}</h2>
      <button
        v-if="isEditable"
        :class="{ _toggled: showSettings }"
        class="header-button"
        @click.stop="handleSettingsClick"
      >
        <v-icon name="ellipsis-h" />
      </button>
    </header>
    <div class="body-wrapper">
      <div :class="bodyClass" class="body" :style="bodyStyle">
        <slot v-if="$slots.default" />
        <template v-else>
          <component
            v-for="(item, idx) in content"
            :is="item.type"
            :key="idx"
            v-bind="item.props"
            class="body-item"
          />
        </template>
      </div>
      <transition name="revealSettings">
        <CellSettings v-if="isEditable && showSettings" />
      </transition>
    </div>
  </div>
</template>

<script>
import merge from 'lodash/merge';
import store from '../../store';
import { createGridTemplate } from '../../utils';
import CellSettings from './CellSettings';

export default {
  name: 'Cell',
  components: {
    CellSettings,
  },
  props: {
    grid: String,
    title: {
      type: String,
      default: '',
    },
    content: {
      type: Array,
      default: () => [],
    },
    padded: {
      type: Boolean,
      default: false,
    },
    editable: {
      type: Boolean,
      default: false,
    },
  },
  created() {
    if (!this.$slots.default) {
      return;
    }
    const { areas } = createGridTemplate(this.grid);
    this.$slots.default.forEach((vnode, idx) => {
      vnode.data = merge(vnode.data, {
        style: {
          gridArea: areas[idx],
        },
      });
    });
  },
  data() {
    return {
      showSettings: false,
    };
  },
  computed: {
    isEditable() {
      return this.editable && store.state.editMode;
    },
    bodyStyle() {
      return {
        gridTemplateAreas: createGridTemplate(this.grid).template,
      };
    },
    bodyClass() {
      return {
        padded: this.padded,
      };
    },
  },
  methods: {
    handleSettingsClick() {
      this.showSettings = !this.showSettings;
    },
  },
};
</script>

<style scoped>
.Cell {
  @apply flex flex-col flex-grow transition duration-300 select-none p-px antialiased;
  background-color: #6277a244;

  &:hover {
    @apply duration-100;
  }
}

.header {
  @apply py-2 px-4 relative flex justify-center items-center;
  @apply uppercase leading-none font-bold;
  color: #fff8;
  font-size: 0.7rem;
}

.title {
  text-shadow: -1px -1px #0001;
}

.header-button {
  @apply absolute top-0 right-0 bottom-0 outline-none z-10 px-4;

  &.focus-visible {
    @apply shadow-outline;
  }

  &._toggled {
    @apply bg-gray-900 text-white;
  }

  &:hover {
    @apply text-white;
  }

  &:active {
    @apply bg-gray-800 transition-none text-white;
  }

  & .fa-icon {
    margin: -0.2em 0;
  }
}

.body-wrapper {
  @apply flex flex-grow relative items-stretch;
}

.body {
  @apply grid gap-px text-gray-400 flex-grow;
}

.body > * {
  @apply bg-gray-900;
}

.padded > * {
  @apply p-2;
}

.revealSettings-enter-active,
.revealSettings-leave-active {
  @apply transition duration-300;
}
.revealSettings-enter,
.revealSettings-leave-to {
  @apply opacity-0;
}
</style>
