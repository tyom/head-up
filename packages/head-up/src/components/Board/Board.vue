<template>
  <div :class="boardClass" class="Board">
    <transition name="slideDown">
      <BoardToolbar
        v-if="!isThumb && isEditing()"
        :board="$props"
      />
    </transition>
    <div :class="layoutClass" class="cells">
      <template v-if="isThumb">
        <div
          v-for="(cell, idx) in cells"
          :key="idx"
          class="cell-placeholder"
        />
      </template>
      <slot v-else-if="!cells.length" />
      <template v-else>
        <Cell
          v-for="(cell, idx) in cells"
          v-bind="cell"
          :editable="editable"
          :key="idx"
          class="cell"
        />
      </template>
    </div>
  </div>
</template>

<script>
import Cell from '../Cell';
import BoardToolbar from './BoardToolbar';

export default {
  name: 'Board',
  components: {
    Cell,
    BoardToolbar,
  },
  props: {
    id: {
      type: String,
      default: '',
    },
    isThumb: {
      type: Boolean,
      default: false,
    },
    title: {
      type: String,
      default: '',
    },
    cells: {
      type: Array,
      default: () => [],
    },
    editable: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    boardClass() {
      return {
        _thumb: this.isThumb,
      };
    },
    layoutClass() {
      const cells = this.$slots.default ? this.$slots.default : this.cells;
      if (!cells.length) {
        return;
      }
      return [`u-grid-${cells.length}-x`];
    },
  },
  inject: ['isEditing'],
};
</script>

<style scoped>
.Board {
  display: flex;
  flex-direction: column;
  flex: 1;
  height: 100%;
}

.cell-placeholder {
  background-color: rgba(#fff, 0.1);
}

.cells {
  height: 100%;
  grid-gap: 0.5em;
  padding: 0.5em;
}

._thumb {
  font-size: 0.8em;
  margin: auto;

  & .cells {
    padding: 2px;
    grid-gap: 2px;
  }

  & .cell {
    border: 0;
    background-color: #fff2;
  }
}

.slideDown-enter-active,
.slideDown-leave-active {
  transition: 0.1s;
  overflow: hidden;
}

.slideDown-enter,
.slideDown-leave-to {
  height: 0;
  opacity: 0;
}
</style>
