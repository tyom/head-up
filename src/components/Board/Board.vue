<template>
  <div :id="boardId" :class="boardClass" class="Board">
    <transition name="slideDown">
      <BoardToolbar
        v-if="!isThumb && editMode"
        :board="$props"
        @save="handleSave"
        @done="handleDone"
      />
    </transition>
    <div :class="layoutClass" class="cells">
      <template v-if="isThumb">
        <div v-for="(cell, idx) in cells" :key="idx" class="cell-placeholder" />
      </template>
      <template v-else>
        <slot v-if="slotCells" />
        <Cell
          v-else
          v-for="(cell, idx) in cells"
          v-bind="cell"
          :editable="editable"
          :key="idx"
          class="cell"
        />
      </template>
    </div>
  </div>
</template>

<script>
import get from 'lodash/get';
import store from '../../store';
import Cell from '../Cell';
import BoardToolbar from './BoardToolbar';

export default {
  name: 'Board',
  components: {
    Cell,
    BoardToolbar,
  },
  props: {
    id: {
      type: String,
      required: true,
    },
    isThumb: {
      type: Boolean,
      default: false,
    },
    title: {
      type: String,
      default: '',
    },
    cells: {
      type: Array,
      default: () => [],
    },
    editable: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    boardId() {
      return this.isThumb ? null : `board-${this.id}`;
    },
    boardClass() {
      return {
        _thumb: this.isThumb,
      };
    },
    editMode() {
      return store.state.editMode;
    },
    slotCells() {
      if (!this.$slots.default) {
        return;
      }
      const cells = this.$slots.default.filter(
        (x) => get(x, 'componentOptions.tag') === 'Cell',
      );
      return cells.length ? cells : undefined;
    },
    layoutClass() {
      const cells = this.slotCells || this.cells;
      if (!cells.length) {
        return;
      }
      return [`u-grid-${cells.length}-x`];
    },
  },
  methods: {
    handleSave(newState) {
      store.dispatch('UPDATE_BOARD', {
        boardId: this.id,
        payload: newState,
      });
    },
    handleDone(newState) {
      if (this.editable) {
        this.handleSave(newState);
      }
      store.dispatch('TOGGLE_EDIT_MODE', false);
    },
  },
};
</script>

<style scoped>
.Board {
  @apply flex flex-grow flex-col h-full;
}

.cell-placeholder {
  background-color: rgba(#fff, 0.1);
}

.cells {
  height: 100%;
  grid-gap: 0.5em;
  padding: 0.5em;
  overflow: auto;
}

._thumb {
  font-size: 0.8em;
  margin: auto;

  & .cells {
    padding: 2px;
    grid-gap: 2px;
  }

  & .cell {
    border: 0;
    background-color: #fff2;
  }
}

.slideDown-enter-active,
.slideDown-leave-active {
  transition: 0.1s;
  overflow: hidden;
}

.slideDown-enter,
.slideDown-leave-to {
  height: 0;
  opacity: 0;
}
</style>
