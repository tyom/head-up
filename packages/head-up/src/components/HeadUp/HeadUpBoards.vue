<template>
  <div class="HeadUpBoards">
    <Board
      v-for="(board, idx) in boards"
      ref="activeBoardEl"
      v-bind="board"
      :editable="!board.isReadOnly"
      :key="board.title + board.id + idx"
    />
    <slot/>
  </div>
</template>

<script>
import Board from '../Board';

export default {
  components: {
    Board,
  },
  props: {
    boards: {
      type: Array,
      default: () => [],
    },
  },
};
</script>

<style scoped>
.HeadUpBoards {
  flex: 1;
  overflow: hidden;
}
</style>
