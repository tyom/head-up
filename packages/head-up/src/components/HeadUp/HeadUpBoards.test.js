import { shallowMount } from '@vue/test-utils';
import Board from '../Board';
import Cell from '../Cell';
import HeadUpBoards from './HeadUpBoards';

const provideMock = {
  isEditing: () => false,
};

const slotContents = `
  <Board title="Slot board #1">
    <Cell title="Cell #1"/>
  </Board>
`;

const boardsMock = [
  {
    title: 'Prop board #1',
    cells: [
      {
        title: 'Cell #1',
      },
      {
        title: 'Cell #2',
      },
    ],
  },
];

test('render with slots', () => {
  const wrapper = shallowMount(HeadUpBoards, {
    slots: {
      default: slotContents,
    },
    stubs: {
      Board,
      Cell,
    },
    provide: provideMock,
  });
  expect(wrapper).toMatchSnapshot();
});

test('render with props', () => {
  const wrapper = shallowMount(HeadUpBoards, {
    propsData: {
      boards: boardsMock,
    },
    stubs: {
      Board,
      Cell,
    },
    provide: provideMock,
  });
  expect(wrapper).toMatchSnapshot();
});

test('render with slot and props', () => {
  const wrapper = shallowMount(HeadUpBoards, {
    propsData: {
      boards: boardsMock,
    },
    slots: {
      default: slotContents,
    },
    stubs: {
      Board,
      Cell,
    },
    provide: provideMock,
  });
  expect(wrapper).toMatchSnapshot();
});
