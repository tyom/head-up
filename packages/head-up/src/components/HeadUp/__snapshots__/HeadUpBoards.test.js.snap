// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render with props 1`] = `
<div class="HeadUpBoards">
  <div class="Board">
    <!---->
    <div class="cells u-grid-2-x">
      <div class="Cell cell">
        <header class="header">
          <h2 class="title">Cell #1</h2>
          <!---->
        </header>
        <div class="container">
          <div class="body"></div>
          <!---->
        </div>
      </div>
      <div class="Cell cell">
        <header class="header">
          <h2 class="title">Cell #2</h2>
          <!---->
        </header>
        <div class="container">
          <div class="body"></div>
          <!---->
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`render with slot and props 1`] = `
<div class="HeadUpBoards">
  <div class="Board">
    <!---->
    <div class="cells u-grid-2-x">
      <div class="Cell cell">
        <header class="header">
          <h2 class="title">Cell #1</h2>
          <!---->
        </header>
        <div class="container">
          <div class="body"></div>
          <!---->
        </div>
      </div>
      <div class="Cell cell">
        <header class="header">
          <h2 class="title">Cell #2</h2>
          <!---->
        </header>
        <div class="container">
          <div class="body"></div>
          <!---->
        </div>
      </div>
    </div>
  </div>
  <div class="Board">
    <!---->
    <div class="cells u-grid-1-x">
      <div class="Cell">
        <header class="header">
          <h2 class="title">Cell #1</h2>
          <!---->
        </header>
        <div class="container">
          <div class="body"></div>
          <!---->
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`render with slots 1`] = `
<div class="HeadUpBoards">
  <div class="Board">
    <!---->
    <div class="cells u-grid-1-x">
      <div class="Cell">
        <header class="header">
          <h2 class="title">Cell #1</h2>
          <!---->
        </header>
        <div class="container">
          <div class="body"></div>
          <!---->
        </div>
      </div>
    </div>
  </div>
</div>
`;
